pkg/guacrest/openapi.yaml

```diff
@@ -28,39 +28,108 @@ paths:
+          $ref: "#/components/responses/BadRequest"
+        "500":
+          $ref: "#/components/responses/InternalServerError"
+        "502":
+          $ref: "#/components/responses/BadGateway"
+  "/v0/package/{purl}":
+    get:
+      summary: Get all purls related to the given purl
+      description: >
+        If a partial purl, pkg:foo/bar is passed in, then it would find all purls 
+        associated with the given purl, such as pkg:foo/bar@1.0, and pkg:foo/bar@2.0.
+      operationId: getPackagePurls
+      parameters:
         - name: purl
-          description: The purl of the dependent package.
-          in: query
-          required: false
+          in: path
+          required: true
+          description: URL-encoded Package URL (purl)
           schema:
             type: string
-        - name: digest
-          description: The digest of the dependent package.
+      responses:
+        "200":
+          $ref: "#/components/responses/PurlList"
+        "400":
+          $ref: "#/components/responses/BadRequest"
+        "500":
+          $ref: "#/components/responses/InternalServerError"
+        "502":
+          $ref: "#/components/responses/BadGateway"
+  "/v0/package/{purl}/vulns":
+    get:
+      summary: Get vulnerabilities for a Package URL (purl)
+      description: >
+        If a partial purl is passed in, then it will find all associated purls 
+        and find vulnerabilities of those. For example, if the partial purl 
+        pkg:foo/bar is passed in, it would find all purls associated with 
+        pkg:foo/bar, such as pkg:foo/bar@1.0, and pkg:foo/bar@2.0. This endpoint 
+        will then find all vulnerabilities for these purls. If the `includeDependencies`
+        flag is set to true, it will also include vulnerabilities of the dependencies.
+      operationId: getPackageVulns
+      parameters:
+        - name: purl
+          in: path
+          required: true
+          description: URL-encoded Package URL (purl)
+          schema:
+            type: string
+        - name: includeDependencies
           in: query
           required: false
+          description: >
+            A flag to include vulnerabilities of the dependencies. If true, the 
+            response will include vulnerabilities for the purl and its dependencies 
+            instead of the vulnerabilities of just the purl.
+          schema:
+            type: boolean
+            default: false
+      responses:
+        "200":
+          $ref: "#/components/responses/VulnerabilityList"
+        "400":
+          $ref: "#/components/responses/BadRequest"
+        "500":
+          $ref: "#/components/responses/InternalServerError"
+        "502":
+          $ref: "#/components/responses/BadGateway"
+  "/v0/package/{purl}/dependencies":
+    get:
+      summary: Get dependencies for a specific Package URL (purl)
+      description: >
+        If a partial purl is passed in, then it will find all associated purls 
```

Comment:
IMO these endpoints should operate on a **single** purl/package that is defined in Guac, instead of handling partial purls. Doing so hides information from the user, and it's not clear if what is returned is directly related to the input package or if it originates from a similar package. Supporting partial purls increases complexity, compounds the inaccuracy of finding the dependencies of a package (https://blog.deps.dev/zillions-of-sboms/), and may lead to drift between the different endpoints.

If a user wants to retrieve the dependencies/vulns of a partial purl, they can just compose `/v0/package/{purl}` and the dependencies endpoint or `/v0/package/{purl}` and the vuln endpoint.

Reply:
I agree, so both the vulnerability and dependency retrievals won't search for partial purls.